zipline/pipeline/domain.py

```diff
@@ -0,0 +1,182 @@
+"""
+Domains
+-------
+
+TODO_SS
+"""
+from interface import implements, Interface
+
+from trading_calendars import get_calendar
+
+from zipline.country import CountryCode
+from zipline.utils.memoize import lazyval
+
+from .sentinels import NotSpecified
+
+
+class IDomain(Interface):
+    """Domain interface.
+    """
+
+    @property
+    def name(self):
+        """User-facing name for this domain.
+        """
+
+    @property
+    def country_code(self):
+        """Country code for assets on this domain.
+        """
+
+    # TODO_SS: The original design for domains was to have them return a
+    # TradingCalendar, but we have a bunch of tests in test_blaze that use very
+    # short session indices that I don't know how to port to using a proper
+    # TradingCalendar.
+    #
+    # Is there a strong reason to prefer just exposing the calendar
+    # vs. exposing the sessions? If so, what do we do about the blaze tests?
+    def get_sessions(self):
+        """Get all trading sessions for the calendar of this domain.
+        """
+
+
+# Create a base class so that we can do type-based validation of user input.
+Domain = implements(IDomain)
+Domain.__name__ = 'Domain'
+
+
+# TODO: Better name for this?
```

Comment:
`CountryDomain`, or somesuch might be more descriptive if we have iterate through subclasses to get a list of domains?

Reply:
CountryDomain seems not quite right because, at least right now, all domains have a country code (that might change in the future, but it will be a pretty significant change if/when it happens). 

Really the difference here is whether we get our dates via a named calendar or via a simple DatetimeIndex.